Trying to support non cli

phyzical opened the pull request:
Fixes #18 
Fixes #20 

## Proposed Changes

  - Added a new cli-index that is exposed on install into bin folder
  - added ability to use require() for nodejs build scripts
  - added constants to assist with typing on what type of dependencies to manipulate
  - added tests
  - slight tweeks to make code more testable
  - added a common run function to allow for NodeJS usage
  - QOL things such as eslint and editor config to enforce conformation between devs :)

Commit:
trying to test non cli

Commit:
addeding a log to test

Commit:
lets try this

Commit:
Trying-To-Support-non-cli: Merge branch 'Trying-To-Support-non-cli' of github.com:phyzical/npm-add-dependencies into Trying-To-Support-non-cli # Please enter a commit message to explain why this merge is necessary, # especially if it merges an updated upstream into a topic branch. # # Lines starting with '#' will be ignored, and an empty message aborts # the commit.

Commit:
lets just try using lib

Commit:
* added the ability to simply call run

Commit:
added readme change

Commit:
one more readme modifcaiton

Commit:
revert removal of new line

Commit:
added constants to support the specification of target from nodejs side

Commit:
readme update

Commit:
fix to how i defined constants

Commit:
* added editorconfig
* added eslintrc
* added packagejson changes to support linting
* QOL changes :)

Commit:
* added a change to attempt to see if we can use require now

Commit:
and the file change itself

Commit:
so it turns out a - breaks npm install haha

Commit:
missed line opps

Commit:
alright?.. lets try a roll back

Commit:
hmm

Commit:
ah.. opps my cheery pick of line borked package json

Commit:
adding inital tests

Commit:
added linting changes

Commit:
* npm ignore updates
* added the clie index again now that package json isnt borked

Commit:
* committing initial tests before testing the removal incase i lose something
* one slight change to AddDeps to support promise trickle down

Commit:
*tear down fixed
* its getting there!

Commit:
final iteration of the index.js top level test

Commit:
* added another refactor to make tests look cleaner
* refactored cli to see if it can use run too

Commit:
added cli mocking inital files
* argv mocked
* figured out how to mock things like process.exit correctly

Commit:
moved around tests for cleaner postioning
various fixes to reduce false positives
tried testing cli failed
started the class level tests

Commit:
disable failing trest

Commit:
Files tests

Commit:
added another return to avoid the race conditions occurring during testing

Commit:
small change to add commit message for draft pr

fixes #18
fixes #20

phyzical commented:
Hey @arfeo,

considering its another working week im only going to get a small amount of time to get the remaining tests done, so figured id get a draft pr going for you to quick scan in case you see anything that makes you too nervous.

I'm still trying to figure out the correct way to test CLI as it seems my argv mock isn't working as intended but i'm sure i will get there soon enough, i have still got the main AddDependencies file to test at the individual function level also.

i tested this in both a cli context and nodejs context before reopening this draft pr in case your wondering `if it still works`.

Hopefully there isn't much that makes you too nervous 👍

Commit:
npmignore rejig

Commit:
readme fix

phyzical commented on index.js:
```diff
@@ -1,11 +1,2 @@
-
-const app = new AddDependencies();
-
-app.addDependencies().then(app.saveToPackage.bind(app)).catch((error) => {
```
this was one change i was unsure of, as to why `app.saveToPackage.bind(app)` was being done?

arfeo replied:
Obviously, because `saveToPackage` method is supposed to use the class context.

phyzical replied:
but why does it not have the class context from within itself already? or was this something todo with the asyncy nature of the package

Commit:
simplification of random filename as its not needed to be so complex

Commit:
unneeded anon func?

arfeo commented:
> Hey @arfeo,
> 
> considering its another working week im only going to get a small amount of time to get the remaining tests done, so figured id get a draft pr going for you to quick scan in case you see anything that makes you too nervous.
> 
> I'm still trying to figure out the correct way to test CLI as it seems my argv mock isn't working as intended but i'm sure i will get there soon enough, i have still got the main AddDependencies file to test at the individual function level also.
> 
> i tested this in both a cli context and nodejs context before reopening this draft pr in case your wondering `if it still works`.
> 
> Hopefully there isn't much that makes you too nervous 👍

You should give me some time. This PR appears to be quite huge.

phyzical commented:
No rush in the slightest 👍  

As i said still have a few more tests to go :)

Commit:
* removal of argv mock
* addition of npm argv mock
* fixed console.logs being shared between tests
* finished majority of tests just cli-index to go
* added a return if no dep passed to runNpmShow to avoid .split error

Commit:
* slight refactors to utils
* tried testing toplevel cli-index again but i think its a lost casue

phyzical commented:
hey @arfeo ! 

finally got some time today to have another crack at the tests. When you get time please give it a review :)

the only set of tests i couldn't seem to get working were the top level cli-index tests, i think this was due to when i try to trigger it with a "require('cli-index')" there's no way for me to tell jest "Okay its done check for expects now please". But this shouldn't be a big deal anyway as im testing the functions themselves also.

so a  quick summary 
* Added a new cli-index that is exposed on install into bin folder
* added ability to use require() for nodejs build scripts
* added constants to assist with typing on what type of dependencies to manipulate
* added tests
* slight tweeks to make code more testable
* added a common run function to allow for NodeJS usage
* QOL things such as eslint and editor config to enforce conformation between devs :)

noftaly commented:
Hey! I'd really like to see this feature added, I also need it.
The scope of the PR seems quite large, it added ESLint, .editorconfig, tests, etc. I think it should be made into separate PRs to be easier to review and to merge only one thing at a time (because it seems that some tests aren't perfect enough, a lot are commented out, and there are some race conditions according to the comments).
Although it might be a little too late 😄

*By the way, multiple files don't have a trailing newline, and you should add ESModules exports*

Anyway is there still a chance for this to be merged? 😄

phyzical commented:
ill split them out over the weekend. i agree the pr is fairly large.

as for the tests they are all working i just couldn't figure out how to test it as a cli module

ill look into resolving point too "By the way, multiple files don't have a trailing newline, and you should add ESModules exports" 👍

phyzical committed:
Trying-To-Support-non-cli: adding linting fixes

phyzical committed:
Trying-To-Support-non-cli: more liniting

phyzical committed:
Trying-To-Support-non-cli: changed to es6 exports, more linting

phyzical committed:
Update README.md

fix to readme for new export

noftaly commented:
Hey! I would just like to clarify: when I suggested using ESModule exports, I was talking about allowing people using native Nodejs' modules (with `"type": "module"` in the package.json) to import it. Using babel is just syntactic sugar, but not actual ESModule exports: it will compile into CommonJS so it doesn't change anything, except for a cleaner code 🙂

To export the library as a native ESModule (in addition to CommonJS and/or babel), you would need to create an `index.mjs` file exporting the class, etc. Example [here](https://github.com/discordjs/discord.js/blob/master/esm/discord.mjs), on the `discord.js` library.

Btw, it might be a bit hard if you've never done TS but I think you should add an `index.d.ts` (Typescript definitions)... (I'm trying to do it if you want, but I've never done TS definitions so I'm not really sure about what I'm doing 😅)
Also, when using it as a dependency, it will display console messages, right? I think it should not 🤔

Otherwise, I'm really looking forward to these changes to be merged 😄

phyzical commented:
@noftaly thanks for the input. im a bit of a scrub when it comes to ESModules (as you can tell haha). i got the tests passing by applying the babel then only just got around to testing it with node just now and yeah it didnt play ball.

im going to roll back what i did and just make a .mjs file as your link suggests to give people the option :+1:  the one thing im a bit vague on at this moment is how  [here](https://github.com/discordjs/discord.js/blob/master/esm/discord.mjs) can call import when they just do module exports in the original index.js, is it magic when `"type": "module"` is provided that it will just work? 
edit: yep its just magic :+1: 

ive done a bit of TS (not much) but it is something ive wanted to learn more (code completion is king imo) so if you don't beat me to it ill add that too :)

phyzical committed:
Trying-To-Support-non-cli: rollback the export change

phyzical committed:
Trying-To-Support-non-cli: Merge branch 'Trying-To-Support-non-cli' of github.com:phyzical/npm-add-dependencies into Trying-To-Support-non-cli # Please enter a commit message to explain why this merge is necessary, # especially if it merges an updated upstream into a topic branch. # # Lines starting with '#' will be ignored, and an empty message aborts # the commit.

phyzical committed:
Trying-To-Support-non-cli: add readme fix

phyzical committed:
Trying-To-Support-non-cli: adding inital typescript funtionailty

phyzical commented:
@noftaly Hey again, i confirmed that the revert + .tsm file works for import with the `type: module` in package.json :+1:  and require still works as it did before

ive also added a script entry for typescript generation for auto .d.ts files ill look into the best way to make the .ts files hopefully tomorrow unless you do it/ find a way :)

noftaly commented:
Great job!
I would have done the exact same thing but in a single file `index.d.ts`, under a namespace (as it is done inside most of DefinitelyTyped's packages). But if this works, then you can let it like this 🙂

arfeo commented:
Hey guys. How're you doing?

phyzical commented:
Hey @arfeo, Hope you have been well! 

This Pr is ready for review when you get a chance 👍 

The only item left to cross off is how to test the cli context of npm-add. i will have one more try at this later today before throwing in the towel on this

phyzical committed:
Trying-To-Support-non-cli: updated the typescripting to support index

phyzical committed:
Trying-To-Support-non-cli: added logic to mimic a cli usage added logic to fix test package.json file names

phyzical committed:
Trying-To-Support-non-cli: fixed a bug in the regex tester not working in cli context

phyzical committed:
Trying-To-Support-non-cli: finished cli context tests

phyzical commented:
one final followup: 

finished the cli context tests so it should be fully ready now :+1:

theoludwig commented:
@arfeo 👀

arfeo commented:
Actually, I would propose to split this huuuuuge PR to a few smaller ones. The current PR is overloaded with new features which have little to nothing to do with its title.

arfeo commented:
Let's start with the main idea of this PR... No jest, no eslint, whatever. Not this time.

LUC18fknU7P commented on lib/AddDependencies.js:
```diff
@@ -38,7 +43,7 @@ class AddDependencies {
             break;
           }
           default: {
-            if (/package\.json/.test(val)) {
```
Shouldn't these kind of changes should be split to separate PRs?

phyzical replied:
yeah i should have, after refactoring this pr out into separate ones i found it wasn't needed.

phyzical commented:
Hey @arfeo,

Sorry i didnt see the last reply.

I have created 3 prs.

https://github.com/arfeo/npm-add-dependencies/pull/23 - node update

https://github.com/arfeo/npm-add-dependencies/pull/24 - test update

https://github.com/arfeo/npm-add-dependencies/pull/25 - typescript update

Linting - TODO once these are ready
